StallMonitor service to detect thread stalls during module processing

knoepfel opened the pull request:
This PR is to resolve issue #15594.  The changes included are:
- a new service called `StallMonitor`, which can output stream-time information to a dedicated file
- a new class called `ThreadSafeOutputFileStream`
- adjustments to the `edmStreamStallGrapher` to use the output of `StallMonitor`

knoepfel committed:
First commit for StallMonitor.

knoepfel committed:
Create and test ThreadSafeOutputFileStream.

knoepfel committed:
Tweak design of StallMonitor.

In addition:
- adjust ThreadSafeOutputFileStream (removing implicit '\n' insertion).
- add Column class used for formatting tabular output.

knoepfel committed:
Switch to using tbb::concurrent_unordered_map.

knoepfel committed:
Implementation/print-out tweaks.

knoepfel committed:
White space changes only.

knoepfel committed:
Adjust graphing script to accommodate StallMonitor output.

knoepfel committed:
Adjust test scripts to use new stall-monitor service.

cmsbuild commented:
A new Pull Request was created by  @knoepfel (Kyle Knoepfel) for CMSSW_9_0_X.

It involves the following packages:

FWCore/Concurrency
FWCore/Services
FWCore/Utilities


@cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @makortel, @wddgit, @wmtan this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

Dr15Jones commented:
Please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
Did `edmStreamStallGrapher.py` retain the ability to parse the output from the Timer Service? I was hoping that could be done.

Dr15Jones commented on FWCore/Utilities/interface/Column.h:
```diff
@@ -0,0 +1,93 @@
+    T t;
+  };
+
+  class Column {
```
I don't think `edm::Column` is an overly descriptive name given it is in such a widely used namespace. I suggest either a more descriptive name `OStreamColumn` or place it into another namespace within `edm::`.

knoepfel replied:
I'm fine with `OStreamColumn`.

knoepfel commented:
No, it didn't.  Not too difficult to reimplement it, though.  Would you prefer the script to "detect" what kind of output it's parsing?  Or should a program option be added for the user to specify which output format they're using?

Dr15Jones commented on FWCore/Concurrency/test/test_threadSafeOutputFileStream.cpp:
```diff
@@ -0,0 +1,31 @@
+namespace {
+  void
+  logToFile(unsigned const thread_index, edm::ThreadSafeOutputFileStream& f)
+  {
```
I suggest adding a synchronization point (e.g. decreasing a global `std::atomic<int>`) so that all threads actually start work at the same time. Else, it is very easy for one thread to do all its work before the next thread actually starts.
```
namespace { std::atomic<int> nWaitingForThreads; }
```
In main
```
nWaitingForThreads = threads.size();
```
in logToFile
```
--nWaitingForThreads;
while( 0 != nWaitingForThreads);
```

knoepfel replied:
Okay, sounds good.

Dr15Jones commented on FWCore/Concurrency/test/streamGrapher_stallMonitor_cfg.py:
```diff
@@ -23,4 +23,4 @@
                                       numberOfThreads = cms.untracked.uint32(5))
 
-process.add_(cms.Service("Tracer", printTimestamps = cms.untracked.bool(True)))
+process.add_(cms.Service("StallMonitor", fileName = cms.untracked.string("stallMonitor.log")))
```
We'd have both a `Tracer` and `StallMonitor` test.

knoepfel replied:
My understanding was that the `StallMonitor` was replacing the `Tracer` for looking for stalls, hence the change.  However, if you'd like to retain both for comparison reasons, I can add the `Tracer` service back.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+      void postEndJob();
+
+      ThreadSafeOutputFileStream file_;
+      bool validFile_; // Separate data member from file to improve efficiency.
```
Can this be `const`?

knoepfel replied:
Yes.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+    // c'tor receiving 'std::string const&' type not provided since we
+    // must be able to call (e.g.) std::vector<StallStatistics>(20),
+    // for which a default label is not sensible in this context.
+
```
How about a `StallStatistics() = default` just to make your intention clear?

knoepfel replied:
If you think that's helpful, I'm fine with that.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+    }
+
+  private:
+    std::string label_ {};
```
Can this be `const`?

knoepfel replied:
No, the `label_` must be set apart from the c'tor for reasons explained in the c'tor comments above...otherwise it could be.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+      auto const thisTime = ms.count();
+      totalTime_ += thisTime;
+      rep_t max {maxTime_};
+      while (thisTime > max && !maxTime_.compare_exchange_weak(max, thisTime)) {}
```
I'm not a big fan of `compare_exchange_weak` especially on Intel hardware where I believe such an instruction doesn't really exist. I suggest always sticking to `compare_exchange_strong` since the semantics are easier to get right.

knoepfel replied:
Ok, `compare_exchange_strong` it is.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+}
+
+namespace {
+  constexpr char const* filename_default {""};
```
Should this be `constexpr char const* const`?

knoepfel replied:
An `auto* const` construct only makes sense in a run-time environment.  Since `filename_default` is a compile-time constant (via `constexpr`), the extra `const` is unnecessary, even if permitted by the compiler.

Dr15Jones replied:
I just tested and the compiler (at least clang) does not allow it.

knoepfel replied:
Wow.  What version of clang?  My old clang version (LLVM 7.0.2) does permit `constexpr char const* const`.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+      auto const& label = moduleLabels_[i];
+      if (label.empty()) continue; // See comment in filling of moduleLabels_;
+      oss << "#M " << std::setw(width) << std::left << col0(i) << space
+          << std::left << moduleLabels_[i] << '\n';
```
It appears `moduleLabels_` is not used beyond this point. Perhaps you could call `moduleLabels_.clear()` to free up some memory?

knoepfel replied:
You're right.  Let's free up the memory.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+  // type containing atomics (like 'StallStatistics')--i.e. atomics
+  // have no copy/move-assignment operators, we must specify the size
+  // of the vector at construction time.
+  moduleStats_ = std::vector<StallStatistics>(moduleLabels_.size());
```
For your case, you could hand write copy constructor for `StallStatistics` since the copy doesn't have to be thread-safe. I'm fine either way.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+
+void StallMonitor::preEvent(StreamContext const& sc)
+{
+  if (!validFile_) return;
```
Since you do not want to do anything if `validFile_` is not needed, I'd suggest not even calling `iRegistry.watchPreEvent`. That will actually improve the efficiency of the job a bit since the framework will not have to call this function each event.

knoepfel replied:
Ah, yes, that's a much better solution.  Will adopt for this instance, and all others you pointed out below.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+  auto const preModEvent = now();
+  auto const sid = stream_id(sc);
+  auto const mid = module_id(mcc);
+  if (file_) {
```
Do you want `validFile_` here?

knoepfel replied:
Yes, that's what I intended.  Good catch.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+
+void StallMonitor::preEventReadFromSource(StreamContext const& sc, ModuleCallingContext const& mcc)
+{
+  if (!validFile_) return;
```
Again, best to not even register this callback.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+
+void StallMonitor::postEventReadFromSource(StreamContext const& sc, ModuleCallingContext const& mcc)
+{
+  if (!validFile_) return;
```
No need to register the callback

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+
+void StallMonitor::postModuleEvent(StreamContext const& sc, ModuleCallingContext const& mcc)
+{
+  if (!validFile_) return;
```
No need to register.

Dr15Jones commented on FWCore/Services/plugins/StallMonitor.cc:
```diff
@@ -0,0 +1,403 @@
+
+void StallMonitor::postEvent(StreamContext const& sc)
+{
+  if (!validFile_) return;
```
No need to register

Dr15Jones commented:
If auto detection of intput file type can be done easily, then I'm all for it.

cmsbuild commented:
-1

Tested at: 7a9eaceee818ec40738f4681439bce3bce537511

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16658/16430/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>      m_taskChosen(ATOMIC_FLAG_INIT),
                   ^~~~~~~~~~~~~~~~
/cvmfs/cms-ib.cern.ch/2016-47/slc6_amd64_gcc530/external/gcc/5.3.0/lib/gcc/x86_64-pc-linux-gnu/5.3.0/../../../../include/c++/5.3.0/bits/atomic_base.h:157:26: note: expanded from macro 'ATOMIC_FLAG_INIT'
#define ATOMIC_FLAG_INIT { 0 }
                         ^~~~~
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-11-16-2300/src/FWCore/Services/plugins/StallMonitor.cc:367:7: error: use of overloaded operator '<<' is ambiguous (with operand types 'edm::LogAbsolute' and 'edm::Column')
  out << tag << space
  ~~~ ^  ~~~
/cvmfs/cms-ib.cern.ch/2016-47/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_9_0_X_2016-11-16-2300/src/FWCore/MessageLogger/interface/MessageLogger.h:357:5: note: candidate function [with T = edm::Column]
    operator<< (T const & t)  { ap << t; return *this; }
    ^
</pre>

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16658/16430/summary.html

knoepfel committed:
Tweaks based on Chris J.'s comments.

Make operator<< for OStreamColumn specific to std::ostream instead of a template.

knoepfel committed:
Adjust graphing script to accommodate both StallMonitor and Tracer outputs.

cmsbuild commented:
Pull request #16658 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

knoepfel commented:
Here are plots made by `edmStreamStallGrapher.py` are attached; they were created by the
 `StallMonitor` and `Tracer` services:
* [stallMonitor.pdf](https://github.com/cms-sw/cmssw/files/600684/stallMonitor.pdf)
* [tracer.pdf](https://github.com/cms-sw/cmssw/files/600685/tracer.pdf)

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @smuzaffar

cmsbuild commented:
-1

Tested at: 0d6e1c408669b2ffcc8e270e588feb63c4b4f254


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/58ce128d7496750d493d079f6feb9963743eda9c
https://github.com/cms-sw/cmssw/commit/81ed056fa6f53604c9dc39020c7c9a202c338fa7
https://github.com/cms-sw/cmssw/commit/068fb037a459e483622d60633bcf8dddd98baeff
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16658/16477/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16658/16477/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16658/16477/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests  RelVals**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testJetMETCorrectionsType1MET had ERRORS


* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
21234.0 step2
<pre>runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D4_GenSimHLBeamSpotFull+DigiFull_2023D4+RecoFullGlobal_2023D4+HARVESTFullGlobal_2023D4/step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D4_GenSimHLBeamSpotFull+DigiFull_2023D4+RecoFullGlobal_2023D4+HARVESTFullGlobal_2023D4.log

</pre>23234.0 step2
<pre>runTheMatrix-results/23234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D5_GenSimHLBeamSpotFull+DigiFull_2023D5+RecoFullGlobal_2023D5+HARVESTFullGlobal_2023D5/step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D5_GenSimHLBeamSpotFull+DigiFull_2023D5+RecoFullGlobal_2023D5+HARVESTFullGlobal_2023D5.log

</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/58ce128d7496750d493d079f6feb9963743eda9c
https://github.com/cms-sw/cmssw/commit/81ed056fa6f53604c9dc39020c7c9a202c338fa7
https://github.com/cms-sw/cmssw/commit/068fb037a459e483622d60633bcf8dddd98baeff
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16658/16477/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16658/16477/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

Dr15Jones commented:
The test failures are not related to this pull request.

Dr15Jones commented:
@davidlange6 ping
